sunscreen_zkp_backend/src/bulletproofs.rs

```diff
@@ -170,56 +186,66 @@ impl BulletproofsCircuit {
                     self.nodes[idx.index()] = Some(input.into());
                 }
                 Operation::Add => {
-                    let (left, right) = query.get_binary_operands(idx)?;
+                    let (left_idx, right_idx) = query.get_binary_operands(idx)?;
 
-                    let left = self.nodes[left.index()]
+                    let left = self.nodes[left_idx.index()]
                         .as_ref()
-                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(left)))
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(left_idx)))
                         .clone();
 
-                    let right = self.nodes[right.index()]
+                    let right = self.nodes[right_idx.index()]
                         .as_ref()
-                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(right)))
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(right_idx)))
                         .clone();
```

Comment:
so is that vec internal to the bulletproof obj? i'm wondering if we can just put an Rc around it and get this behavior for free. that way even these clones are just bumping reference counts:
```
                    let right = self.nodes[right_idx.index()]
                        .as_ref()
                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(right_idx)))
                        .clone();
```

Reply:
It's not immediately obvious to me how you would do this. What data structure would receive the Rc that doesn't need to be explicitly dropped itself?

Reply:
I was thinking something like
```rust
enum Node {
    LinearCombination(Rc<LinearCombination>),
    Scalar(Scalar),
}
```
that way any large variants of the nodes don't blow up as they get cloned. I just played around with this and got it to compile, but I'm not sure it makes any sense. It requires you to `lc.as_ref().clone()` whenever you want to use the `Node::LInearCombination(lc)` in a constraint.